File: mediumSpringWebflux/src/main/java/com/example/mediumspringwebflux/comment/service/DeleteComment.java
package com.example.mediumspringwebflux.comment.service;

import com.example.mediumspringwebflux.board.repository.BoardRepository;
import com.example.mediumspringwebflux.board.service.facade.FindBoard;
import com.example.mediumspringwebflux.comment.repository.CommentRepository;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
public class DeleteComment {
    private final CommentRepository commentRepository;
    private final BoardRepository boardRepository;
    private final FindBoard findBoard;

    public DeleteComment(CommentRepository commentRepository, BoardRepository boardRepository, FindBoard findBoard) {
        this.commentRepository = commentRepository;
        this.boardRepository = boardRepository;
        this.findBoard = findBoard;
    }

    public Mono<Void> execute(String commentId) {
        return commentRepository.findById(commentId)
                .flatMap(comment ->
                        findBoard.execute(comment.getBoardId())
                                .flatMap(board -> board.deleteComment(comment)
                                        .then(boardRepository.save(board)))
                ).then(commentRepository.deleteById(commentId));
    }

}

File: mediumSpringWebMVC/src/main/kotlin/com/example/mediumspringwebmvc/domain/comment/service/DeleteComment.kt
package com.example.mediumspringwebmvc.domain.comment.service

import com.example.mediumspringwebmvc.domain.comment.repository.CommentRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class DeleteComment(
    private val commentRepository: CommentRepository
) {

    @Transactional
    fun execute(id: Long) {
        val comment = commentRepository.findByIdOrNull(id) ?: throw RuntimeException("댓글을 찾지 못함")

        commentRepository.delete(comment)
    }

}